fms/modules/attention.py

```diff
@@ -58,15 +58,19 @@ def __init__(
         self.emb_v_per_head = emb_v
         self.p_dropout = p_dropout if p_dropout is not None else 0.0
         self.use_bias = use_bias
-        self.query = nn.Linear(
-            self.emb_dim, self.nheads * self.emb_kq_per_head, bias=use_bias
-        )
-        self.key = nn.Linear(
-            self.emb_dim, self.kvheads * self.emb_kq_per_head, bias=use_bias
-        )
-        self.value = nn.Linear(
-            self.emb_dim, self.kvheads * self.emb_v_per_head, bias=use_bias
+
```

Comment:
Will we need some way to load state dicts that were taken from before this change? The model linsong+davis have been training is one of these fms models, and we'll still need to be able to load that I think.

I beleive it would also be possible to retain (non-trainable) attributes of query.weight, key.weight, etc, where changes to any of them trigger an update to the actual qkv_fused? Doing it that way might allow preserving the same state dict conversions, though avoiding memory duplication could make it a messy implementation or special class.

Is the behavior identical in all cases? I know the existing implementation allows for, for example, a kq length that differs from the value length. Do any of the existing parameters become redundant now that the attributes are fused?

Reply:
Ah that is a good question. There are a few different routes we can take:

1. create an adapter for fms unfused attn (something like model_source="fms_unfused_attn"). I am not sure I love this idea as any time we make changes that break our weights, we will end up needing to do this, and it can get messy.
2. I briefly remember looking at overriding `load_state_dict`, not sure if thats an option, but can explore further.
3. The option you provided may work, I would need to try it out and see what that would look like. Do you know of anyone that has done that in the past to fix this backwards compatibility issue of the weights?

> Is the behavior identical in all cases? I know the existing implementation allows for, for example, a kq length that differs from the value length. Do any of the existing parameters become redundant now that the attributes are fused?

Not sure if the params become redundant, however I don't believe with this fused we have kq/value lengths differing. Has this ever been the case?

Reply:
In encoder-decoder models, the lengths of q, k, and v can and will differ

Reply:
This has been fixed in latest commit with a pre-processing step (adapter for handling weights)